File: packages/expo/src/plugins/last-login-method.ts
import type { Awaitable, BetterAuthClientPlugin } from "@better-auth/core";

export interface LastLoginMethodClientConfig {
	storage: {
		setItem: (key: string, value: string) => any;
		getItem: (key: string) => string | null;
		deleteItemAsync: (key: string) => Awaitable<void>;
	};
	/**
	 * Prefix for local storage keys (e.g., "my-app_last_login_method")
	 * @default "better-auth"
	 */
	storagePrefix?: string | undefined;
	/**
	 * Custom resolve method for retrieving the last login method
	 */
	customResolveMethod?:
		| ((url: string | URL) => Awaitable<string | undefined | null>)
		| undefined;
}

const paths = [
	"/callback/",
	"/oauth2/callback/",
	"/sign-in/email",
	"/sign-up/email",
];
const defaultResolveMethod = (url: string | URL) => {
	const { pathname } = new URL(url.toString(), "http://localhost");

	if (paths.some((p) => pathname.includes(p))) {
		return pathname.split("/").pop();
	}
	if (pathname.includes("siwe")) return "siwe";
	if (pathname.includes("/passkey/verify-authentication")) {
		return "passkey";
	}

	return;
};

export const lastLoginMethodClient = (config: LastLoginMethodClientConfig) => {
	const resolveMethod = config.customResolveMethod || defaultResolveMethod;
	const storagePrefix = config.storagePrefix || "better-auth";
	const lastLoginMethodName = `${storagePrefix}_last_login_method`;
	const storage = config.storage;

	return {
		id: "last-login-method-expo",
		fetchPlugins: [
			{
				id: "last-login-method-expo",
				name: "Last Login Method",
				hooks: {
					onResponse: async (ctx) => {
						const lastMethod = await resolveMethod(ctx.request.url);
						if (!lastMethod) {
							return;
						}

						await storage.setItem(lastLoginMethodName, lastMethod);
					},
				},
			},
		],
		getActions() {
			return {
				/**
				 * Get the last used login method from storage
				 *
				 * @returns The last used login method or null if not found
				 */
				getLastUsedLoginMethod: (): string | null => {
					return storage.getItem(lastLoginMethodName);
				},
				/**
				 * Clear the last used login method from storage
				 */
				clearLastUsedLoginMethod: async () => {
					await storage.deleteItemAsync(lastLoginMethodName);
				},
				/**
				 * Check if a specific login method was the last used
				 * @param method The method to check
				 * @returns True if the method was the last used, false otherwise
				 */
				isLastUsedLoginMethod: (method: string): boolean => {
					const lastMethod = storage.getItem(lastLoginMethodName);
					return lastMethod === method;
				},
			};
		},
	} satisfies BetterAuthClientPlugin;
};

File: packages/flutter/dart/lib/src/plugins/last_login_method.dart
import '../storage.dart';
import 'plugin.dart';

/// Options for [lastLoginMethodClient].
final class LastLoginMethodOptions {
	const LastLoginMethodOptions({
		required this.storage,
		this.storagePrefix = 'better-auth',
		this.resolveMethod,
	});

	final AuthStorage storage;
	final String storagePrefix;

	/// Override how a request URL maps to a login-method label.
	final String? Function(Uri url)? resolveMethod;
}

/// Persists the last successful login method (email, google, passkey, …).
final class LastLoginMethodPlugin extends AuthClientPlugin {
	LastLoginMethodPlugin(this.options);

	final LastLoginMethodOptions options;

	@override
	String get id => 'last-login-method';

	String get _key => '${options.storagePrefix}_last_login_method';

	static const _tracked = [
		'/callback/',
		'/oauth2/callback/',
		'/sign-in/email',
		'/sign-up/email',
	];

	String? _defaultResolve(Uri url) {
		final path = url.path;
		if (_tracked.any(path.contains)) {
			final segments = path.split('/').where((s) => s.isNotEmpty);
			return segments.isEmpty ? null : segments.last;
		}
		if (path.contains('siwe')) return 'siwe';
		if (path.contains('/passkey/verify-authentication')) return 'passkey';
		if (path.contains('/sign-in/social')) return 'social';
		return null;
	}

	@override
	Future<void> handleResponse({
		required Uri url,
		required Map<String, String> headers,
		required int statusCode,
		Object? body,
	}) async {
		if (statusCode >= 400) return;
		final resolve = options.resolveMethod ?? _defaultResolve;
		final method = resolve(url);
		if (method == null || method.isEmpty) return;
		await options.storage.setItem(_key, method);
	}

	Future<String?> getLastUsedLoginMethod() => options.storage.getItem(_key);

	Future<void> clearLastUsedLoginMethod() => options.storage.removeItem(_key);

	Future<bool> isLastUsedLoginMethod(String method) async {
		final last = await getLastUsedLoginMethod();
		return last == method;
	}
}

LastLoginMethodPlugin lastLoginMethodClient(LastLoginMethodOptions options) {
	return LastLoginMethodPlugin(options);
}
